packages/flutter/lib/src/material/ink_splash.dart

```diff
@@ -158,14 +161,33 @@ class InkSplash extends InkFeature {
     if (originOffset == null) {
       canvas.save();
       canvas.transform(transform.storage);
-      if (_clipCallback != null)
-        canvas.clipRect(_clipCallback());
+      if (_clipCallback != null) {
+        if (_borderRadius != BorderRadius.zero) {
+          final RRect clipRRect = new RRect.fromRectAndCorners(
+            _clipCallback(),
+            topLeft: _borderRadius.topLeft, topRight: _borderRadius.topRight,
+            bottomLeft: _borderRadius.bottomLeft, bottomRight: _borderRadius.bottomRight,
+          );
+          canvas.clipRRect(clipRRect);
+        } else {
+          canvas.clipRect(_clipCallback());
+        }
+      }
       canvas.drawCircle(center, _radius.value, paint);
       canvas.restore();
     } else {
       if (_clipCallback != null) {
         canvas.save();
-        canvas.clipRect(_clipCallback().shift(originOffset));
+        if (_borderRadius != BorderRadius.zero) {
+          final RRect clipRRect = new RRect.fromRectAndCorners(
+            _clipCallback(),
+            topLeft: _borderRadius.topLeft, topRight: _borderRadius.topRight,
+            bottomLeft: _borderRadius.bottomLeft, bottomRight: _borderRadius.bottomRight,
+          );
+          canvas.clipRRect(clipRRect.shift(originOffset));
+        } else {
+          canvas.clipRect(_clipCallback().shift(originOffset));
+        }
```

Comment:
this is a lot of code to duplicate; can you factor out the code that calls _clipCallback?